tools/profiling/Controller.php

```diff
@@ -252,508 +80,63 @@ public function run()
-        echo '</table>
-		</div>';
-    }
-
-    protected function displayProfilingObjectModel()
-    {
-        echo '
-		<div class="row">
-			<h2><a name="objectModels">ObjectModel instances</a></h2>
-			<table class="table table-condensed">
-				<tr><th>Name</th><th>Instances</th><th>Source</th></tr>';
-        foreach (ObjectModel::$debug_list as $class => $info) {
-            echo '<tr>
-					<td>'.$class.'</td>
-					<td><span '.$this->getObjectModelColor(count($info)).'>'.count($info).'</span></td>
-					<td>';
-            foreach ($info as $trace) {
-                echo str_replace(array(_PS_ROOT_DIR_, '\\'), array('', '/'), $trace['file']).' ['.$trace['line'].']<br />';
-            }
-            echo '	</td>
-				</tr>';
-        }
-        echo '</table>
-		</div>';
-    }
-
-    protected function displayProfilingFiles()
-    {
-        $i = 0;
-
-        echo '<div class="row">
-		<h2><a name="includedFiles">Included Files</a></h2>
-		<table class="table table-condensed">
-			<tr><th>#</th><th>Filename</th></tr>';
-        foreach (get_included_files() as $file) {
-            $file = str_replace('\\', '/', str_replace(_PS_ROOT_DIR_, '', $file));
-            if (strpos($file, '/tools/profiling/') === 0) {
-                continue;
-            }
-            echo '<tr><td>'.(++$i).'</td><td>'.$file.'</td></tr>';
-        }
-        echo '</table>
-		</div>';
-    }
-
-    public function displayProfiling()
+    /**
+     * Display profiling
+     * If it's a migrated page, we change the outPutHtml content, otherwise
+     * we display the profiling at the end of the page.
+     *
+     * @return string
+     */
+    public function displayProfiling(): string
     {
+        $content = '';
         if (!empty($this->redirect_after)) {
-            echo '
-			<html>
-				<head>
-					<meta charset="utf-8" />
-					<script src="//code.jquery.com/jquery-1.11.2.min.js"></script>
-					<script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.2/js/bootstrap.min.js"></script>
-					<link href="//maxcdn.bootstrapcdn.com/bootstrap/3.3.2/css/bootstrap.min.css" rel="stylesheet" type="text/css"/>
-				</head>
-				<body>
-					<div class="container" style="width:100%">
-						<div class="row">
-							<div class="col-lg-12">
-								<h2>Caught redirection to <a href="'.htmlspecialchars($this->redirect_after).'"> '.htmlspecialchars($this->redirect_after).' </a></h2>
-							</div>
-						</div>';
+            $this->context->smarty->assign(
+                [
+                    'redirectAfter' => $this->redirect_after,
+                ]
+            );
+            $content .= $this->context->smarty->fetch(__DIR__ . '/templates/redirect.tpl');
         } else {
             // Call original display method
+            ob_start();
             $this->display();
-            $this->profiler[] = $this->stamp('display');
+            $displayOutput = ob_get_clean();
+            if (empty($displayOutput) && isset($this->outPutHtml)) {
```

Comment:
```suggestion
            if (empty($displayOutput) && isset($this->outputHtml)) {
```

Reply:
Not, the original name is `outPutHtml` :sweat_smile:

Reply:
:(